game-samples/src/golang/matchmaking-service/models/games.go

```diff
@@ -0,0 +1,281 @@
+package models
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math/rand"
+	"time"
+
+	spanner "cloud.google.com/go/spanner"
+	"github.com/google/uuid"
+	iterator "google.golang.org/api/iterator"
+)
+
+type Game struct {
+	GameUUID string   `json:"gameUUID"`
+	Players  []string `json:"players"`
+	Winner   string   `json:"winner"`
+	created  time.Time
+	finished time.Time
+}
+
+type PlayerStats struct {
+	Games_played int `json:"games_played"`
+	Games_won    int `json:"games_won"`
+}
+
+type Player struct {
+	PlayerUUID string           `json:"playerUUID"`
+	Stats      spanner.NullJSON `json:"stats"`
+}
+
+func generateUUID() string {
+	return uuid.NewString()
+}
+
+// Helper function to read rows from a spanner.
+func readRows(iter *spanner.RowIterator) ([]spanner.Row, error) {
+	var rows []spanner.Row
+	defer iter.Stop()
+
+	for {
+		row, err := iter.Next()
+		if err == iterator.Done {
+			break
+		}
+
+		if err != nil {
+			return rows, err
+		}
+
+		rows = append(rows, *row)
+	}
+
+	return rows, nil
+}
+
+// Get players for a game
+// We only care about the playerUUID and their stats, as this is intended to be used
+// to modify players when a game is closed
+func getGamePlayers(gameUUID string, ctx context.Context, txn *spanner.ReadWriteTransaction) ([]string, []Player, error) {
+	stmt := spanner.Statement{
+		SQL: `SELECT PlayerUUID, Stats FROM players
+				INNER JOIN (
+				SELECT pUUID FROM games g, UNNEST(g.Players) AS pUUID WHERE gameUUID=@game
+				) AS gPlayers ON gPlayers.pUUID = players.PlayerUUID;`,
```

Comment:
I think this will be very inefficient as the database grows. You are doing an inner join on a calculated, unindexed column (`gPlayers.pUUID`). I think you should consider redesigning that part of the data model to keep it more performant.
Hmm... reconsidering my initial comment on this. The sub select will of course always return a relatively small number of rows, so it should be possible to come up with a query plan that does the inner select first, and then just filters the larger `players` table based on the results from that. But I think it would be good to verify that it actually does that.

Reply:
thanks for this callout. The reason I went with this data model is to show the flexibility of the ARRAY data type to keep a point-in-time snapshot of game players without creating a brand new relation (`games_players`). I'll definitely check out the execution plans on this, and in the worst case it will be good content to publish "here's how you debug something like this"